API Ergonomics: Consider reorganizing lifetimes in rsfbclient main crate traits/structs

The lifetime in `Queryable<'a, R>` (and `Transaction<'a,C>`, `StmtIter<'a, R,C>` etc)  will tend to infect whatever uses it.
Unfortunately, due to some warts of the rust language, this leads to unexpected compilation errors in some cases.
For example:
```
  const GTT_QRY: &str =
  "select rdb$relation_name
     from rdb$relations
    where     rdb$relation_name = 'GTT_TABLE'
          and rdb$relation_type > 3;";

  const GTT_CREATE: &str =
    "create global temporary table GTT_TABLE
     (table_id INT not null primary key)
     on commit preserve rows;";

  pub fn with_temporary_table<'a, E>( conn: &'a mut E ) -> ()
    where 
      E: Execute + Queryable<'a, (String,)>
  {
    let result : Vec<(String,)> = conn.query(GTT_QRY, ()).unwrap();
    let result2 =  conn.execute(GTT_CREATE, ());
  }

error[E0499]: cannot borrow `*conn` as mutable more than once at a time
   --> src/main.rs:100:20
    |
95  |   pub fn with_temporary_table<'a, E>( conn: &'a mut E ) -> ()
    |                               -- lifetime `'a` defined here
...
99  |     let result : Vec<(String,)> = conn.query(GTT_QRY, ()).unwrap();
    |                                   -----------------------
    |                                   |
    |                                   first mutable borrow occurs here
    |                                   argument requires that `*conn` is borrowed for `'a`
100 |     let result2 =  conn.execute(GTT_CREATE, ());
    |                    ^^^^ second mutable borrow occurs here
```
pardon if i'm misunderstanding something, but I also also don't see why R of `Queryable<'a, R>` should be bounded by `'a` at the trait level like it is in its where clause either:  `where R : FromRow + 'a`
It seems like the lifetime of R should be unrelated to the connection/transaction/iterator at the trait level: Once we have an R on hand, we should be able to do whatever we want with it, or the Connection, in either order, if our impl allows it.

--- a/src/connection/mod.rs
+++ b/src/connection/mod.rs
@@ -446,17 +446,21 @@ where
     }
 }
 
-impl<'a, R, C> Queryable<'a, R> for Connection<C>
+impl<R, C> Queryable<R> for Connection<C>
 where
-    R: FromRow + 'a,
-    C: FirebirdClient + 'a,
+    R: FromRow + 'static,
+    C: FirebirdClient,
 {
-    type Iter = StmtIter<'a, R, C>;
+    // type Iter = StmtIter<'a, R, C>;
 
     /// Prepare, execute, return the rows and commit the sql query
     ///
     /// Use `()` for no parameters or a tuple of parameters
-    fn query_iter<P>(&'a mut self, sql: &str, params: P) -> Result<Self::Iter, FbError>
+    fn query_iter<'a, P>(
+        &'a mut self,
+        sql: &str,
+        params: P,
+    ) -> Result<Box<dyn Iterator<Item = Result<R, FbError>> + 'a>, FbError>
     where
         P: IntoParams,
     {
@@ -476,7 +480,7 @@ where
                     _marker: Default::default(),
                 };
 
-                Ok(iter)
+                Ok(Box::new(iter))
             }
             Err(e) => {
                 // Return the statement to the cache

--- a/src/connection/mod.rs
+++ b/src/connection/mod.rs
@@ -450,8 +450,6 @@ impl<C> Queryable for Connection<C>
 where
     C: FirebirdClient,
 {
-    // type Iter = StmtIter<'a, R, C>;
-
     /// Prepare, execute, return the rows and commit the sql query
     ///
     /// Use `()` for no parameters or a tuple of parameters

--- a/src/query.rs
+++ b/src/query.rs
@@ -7,23 +7,25 @@
 use rsfbclient_core::{FbError, FromRow, IntoParams};
 
 /// Implemented for types that can be used to execute sql queries
-pub trait Queryable<'a, R>
+pub trait Queryable<R>
 where
-    R: FromRow + 'a,
+    R: FromRow,
 {
-    type Iter: Iterator<Item = Result<R, FbError>> + 'a;
-
     /// Returns the results of the query as an iterator
     ///
     /// Use `()` for no parameters or a tuple of parameters
-    fn query_iter<P>(&'a mut self, sql: &str, params: P) -> Result<Self::Iter, FbError>
+    fn query_iter<'a, P>(
+        &'a mut self,
+        sql: &str,
+        params: P,
+    ) -> Result<Box<dyn Iterator<Item = Result<R, FbError>> + 'a>, FbError>
     where
         P: IntoParams;
 
     /// Returns the results of the query as a `Vec`
     ///
     /// Use `()` for no parameters or a tuple of parameters
-    fn query<P>(&'a mut self, sql: &str, params: P) -> Result<Vec<R>, FbError>
+    fn query<'a, P>(&'a mut self, sql: &str, params: P) -> Result<Vec<R>, FbError>
     where
         P: IntoParams,
     {
@@ -33,7 +35,7 @@ where
     /// Returns the first result of the query, or None
     ///
     /// Use `()` for no parameters or a tuple of parameters
-    fn query_first<P>(&'a mut self, sql: &str, params: P) -> Result<Option<R>, FbError>
+    fn query_first<'a, P>(&'a mut self, sql: &str, params: P) -> Result<Option<R>, FbError>
     where
         P: IntoParams,
     {

--- a/src/transaction.rs
+++ b/src/transaction.rs
@@ -128,8 +128,6 @@ impl<'c, C> Queryable for Transaction<'c, C>
 where
     C: FirebirdClient,
 {
-    // type Iter = StmtIter<'a, R, C>;
-
     /// Prepare, execute and return the rows of the sql query
     ///
     /// Use `()` for no parameters or a tuple of parameters

--- a/src/transaction.rs
+++ b/src/transaction.rs
@@ -124,17 +124,21 @@ where
     }
 }
 
-impl<'a, R, C> Queryable<'a, R> for Transaction<'a, C>
+impl<'c, R, C> Queryable<R> for Transaction<'c, C>
 where
-    R: FromRow + 'a,
+    R: FromRow + 'c,
     C: FirebirdClient,
 {
-    type Iter = StmtIter<'a, R, C>;
+    // type Iter = StmtIter<'a, R, C>;
 
     /// Prepare, execute and return the rows of the sql query
     ///
     /// Use `()` for no parameters or a tuple of parameters
-    fn query_iter<P>(&'a mut self, sql: &str, params: P) -> Result<Self::Iter, FbError>
+    fn query_iter<'a, P>(
+        &'a mut self,
+        sql: &str,
+        params: P,
+    ) -> Result<Box<dyn Iterator<Item = Result<R, FbError>> + 'a>, FbError>
     where
         P: IntoParams,
     {
@@ -156,7 +160,7 @@ where
                     _marker: Default::default(),
                 };
 
-                Ok(iter)
+                Ok(Box::new(iter))
             }
             Err(e) => {
                 // Return the statement to the cache

